segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java

```diff
@@ -164,12 +164,15 @@ public TableBasedMetadataStore(String tableName, TableStore tableStore, Executor
                             }, getExecutor())
                             .thenComposeAsync(v2 -> {
                                 // Delete deleted keys.
-                                return this.tableStore.remove(tableName, keysToDelete, timeout);
+                                return this.tableStore.remove(tableName, keysToDelete, timeout)
+                                        .handleAsync((v1, ex) -> {
+                                            if (ex == null) {
+                                                deletedKeyToTxnDataMap.values().stream().forEach(txnData -> txnData.setDbObject(TableKey.NOT_EXISTS));
+                                            }
+                                            return null;
+                                        }, getExecutor());
                             }, getExecutor())
                             .thenRunAsync(() -> {
-                                for (val deletedKey : keysToDelete) {
-                                    deletedKeyToTxnDataMap.get(deletedKey).setDbObject(TableKey.NOT_EXISTS);
-                                }
                                 TABLE_WRITE_LATENCY.reportSuccessEvent(t.getElapsed());
```

Comment:
Can we move this into the `handleAsync` callback above so we don't create a task just to report a metric?

Reply:
simplified flow now.